core/src/crds_trie.rs

```diff
@@ -0,0 +1,359 @@
+        self.remove_node(
+            CrdsFilter::hash_as_u64(&value.value_hash),
+            &value.value.label(),
+        )
+    }
+
+    /// Returns all the values where the mask_bits most significant bits of
+    /// their hash is the same as the given bit-mask.
+    pub fn find(&self, mask: u64, mask_bits: u32) -> impl Iterator<Item = &CrdsValueLabel> {
+        let node = self.find_node(mask, mask_bits);
+        CrdsTrieIterator::new(node).map(|v| v.iter()).flatten()
+    }
+
+    /// Traverses the trie following the bit pattern of the given hash value
+    /// and inserts the value at the final leaf node. Returns false if the trie
+    /// does not satisfy required invariants (which should indicate a bug).
+    fn insert_node(&mut self, hash: u64, value: CrdsValueLabel) -> bool {
+        let mut node = self;
+        let mut bit = MSB_U64;
+        while bit != 0 {
+            let branch = match node {
+                CrdsTrie::Leaf(_) => return false,
+                CrdsTrie::Node(left, size, right) => {
+                    *size += 1;
+                    if bit & hash == 0 {
+                        left
+                    } else {
+                        right
+                    }
+                }
+            };
+            // Initialize the branch if needed.
+            if branch.is_none() {
+                *branch = new_link(bit == 1);
+            }
+            node = branch.as_mut().unwrap();
+            bit >>= 1;
+        }
+        match node {
+            CrdsTrie::Node(_, _, _) => false,
+            CrdsTrie::Leaf(values) => {
+                values.push(value);
+                true
+            }
+        }
+    }
+
+    /// Traverses the trie following the bit pattern of the given hash value
+    /// and removes the value at the final leaf node. Returns false if the
+    /// value is not found.
+    fn remove_node(&mut self, hash: u64, value: &CrdsValueLabel) -> bool {
+        if self.remove_noprune(hash, &value) {
+            self.prune(hash);
+            true
+        } else {
+            false
+        }
+    }
+
+    fn remove_noprune(&mut self, hash: u64, value: &CrdsValueLabel) -> bool {
+        let mut node = self;
+        let mut bit = MSB_U64;
+        while bit != 0 {
+            let branch = match node {
+                CrdsTrie::Leaf(_) => return false,
+                CrdsTrie::Node(left, _, right) => {
+                    if bit & hash == 0 {
+                        left
+                    } else {
+                        right
+                    }
+                }
+            };
+            node = match branch {
+                None => return false,
+                Some(node) => node,
+            };
+            bit >>= 1;
+        }
+        if let CrdsTrie::Leaf(values) = node {
+            if let Some(index) = values.iter().position(|v| v == value) {
+                values.swap_remove(index);
```

Comment:
So once this remove happens, the "size" field on each of the parent nodes will be inconsistent unless we also call "prune" right? 

I think it would be easier to track to make sure the data structure remains consistent after each function call. For instance, if we keep track of the parents in this search, then after this swap we can just decrement the count in each of the parents and then prune the earliest parent that has `*count == 0`

Reply:
That is a correct observation, but first, these are internal methods not public api, and second the pruning is only an internal optimization. The data structure is still valid even if we do not call into `prune` method.
I did spend some time trying alternatives but the problem is that we need _mutable_ references to the parents along the path to the leaf, and that forces multiple concurrent mutable references which is a compile error.

We can come back and revise this code later, but I think for the first version factoring out `remove` and `prune` logic into separate methods gives a much simpler code.

Reply:
Ah yeah, to get around the mutable references you would need to use `RefCell` and interior mutability.

Sounds good to me for now!